pkg/rotation/reconciler.go

```diff
@@ -0,0 +1,346 @@
+	spcName, spcNamespace := spcps.Status.SecretProviderClassName, spcps.Namespace
+
+	// get the secret provider class the pod status is referencing from informer cache
+	spc, err := r.store.GetSecretProviderClass(spcName, spcNamespace)
+	if err != nil {
+		return fmt.Errorf("failed to get secret provider class %s/%s, err: %+v", spcNamespace, spcName, err)
+	}
+	paramsJSON, err := json.Marshal(spc.Spec.Parameters)
+	if err != nil {
+		return fmt.Errorf("failed to marshal parameters, err: %+v", err)
+	}
+	permissionJSON, err := json.Marshal(permission)
+	if err != nil {
+		return fmt.Errorf("failed to marshal permission, err: %+v", err)
+	}
+	// get pod from informer cache
+	podName, podNamespace := spcps.Status.PodName, spcps.Namespace
+	pod, err := r.store.GetPod(podName, podNamespace)
+	if err != nil {
+		return fmt.Errorf("failed to get pod %s/%s, err: %+v", podNamespace, podName, err)
+	}
+
+	// check if the volume pertaining to the current spc is using nodePublishSecretRef for
+	// accessing external secrets store
+	var nodePublishSecretRef *v1.LocalObjectReference
+	for _, vol := range pod.Spec.Volumes {
+		if vol.CSI.Driver != "secrets-store.csi.k8s.io" {
+			continue
+		}
+		if vol.CSI.VolumeAttributes["secretProviderClass"] != spc.Name {
+			continue
+		}
+		nodePublishSecretRef = vol.CSI.NodePublishSecretRef
+		break
+	}
+
+	var secretsJSON []byte
+	// read the Kubernetes secret referenced in NodePublishSecretRef and marshal it
+	// This comprises the secret parameter in the MountRequest to the provider
+	if nodePublishSecretRef != nil {
+		secretName := nodePublishSecretRef.Name
+		secretNamespace := spcps.Namespace
+
+		// read secret from the informer cache
+		secret, err := r.store.GetSecret(secretName, secretNamespace)
+		if err != nil {
+			return fmt.Errorf("failed to get node publish secret %s/%s, err: %+v", secretNamespace, secretName, err)
+		}
+
+		nodePublishSecretData := make(map[string]string)
+		for k, v := range secret.Data {
+			nodePublishSecretData[k] = string(v)
+		}
+		secretsJSON, err = json.Marshal(nodePublishSecretData)
+		if err != nil {
+			return fmt.Errorf("failed to marshal node publish secret data, err: %+v", err)
+		}
+	}
+
+	// generate a map with the current object versions stored in spc pod status
+	// the old object versions are passed on to the provider as part of the MountRequest.
+	// the provider can use these current object versions to decide if any action is required
+	// and if the objects need to be rotated
+	oldObjectVersions := make(map[string]string)
+	for _, obj := range spcps.Status.Objects {
+		oldObjectVersions[obj.ID] = obj.Version
+	}
+
+	providerClient, err := r.getProviderClient(string(spc.Spec.Provider))
+	if err != nil {
+		return fmt.Errorf("failed to create provider client, err: %+v", err)
+	}
+	newObjectVersions, _, err := providerClient.MountContent(ctx, string(paramsJSON), string(secretsJSON), spcps.Status.TargetPath, string(permissionJSON), oldObjectVersions)
```

Comment:
After the new object version has been mounted, if something fails before the spcps is updated, then what is the mechanism that triggers the reconcile again? the interval? Right now I see any error returned from this function gets logged, can we requeue as well so another reconcile loop gets triggered?

Reply:
Yeah, right now it's controlled by the interval.

Reply:
on error, can we requeue as well so another reconcile loop gets triggered?

Reply:
[processNextItem](https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/303/files#diff-1c6cad8993ddd7d5baf508039d01ff3bR382-R401) picks up the object from the worker queue and reconciles it. In case of errors we [handleError](https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/303/files#diff-1c6cad8993ddd7d5baf508039d01ff3bR403-R410) which requeues the spcps for reconcile after 10s.